snap/quota/quota.go

```diff
@@ -271,11 +272,53 @@ func (grp *Group) getAllowedCPUs() []int {
 	return grp.CPULimit.AllowedCPUs
 }
 
+// GetCPUSetQuota returns the currently active CPU set quota for this group, which
+// could originate from a parent
+func (grp *Group) GetCPUSetQuota() []int {
+	if grp.CPULimit != nil && len(grp.CPULimit.AllowedCPUs) != 0 {
+		return grp.CPULimit.AllowedCPUs
+	}
+
+	parent := grp.parentGroup
+	for parent != nil {
+		if parent.CPULimit != nil && len(parent.CPULimit.AllowedCPUs) != 0 {
+			return parent.CPULimit.AllowedCPUs
+		}
+		parent = parent.parentGroup
+	}
+	return nil
+}
+
+// export it for test
+var runtimeNumCPU = runtime.NumCPU
+
+// GetCorrectedCPUCount returns the maximum number of allowed CPU cores for
+// this group. It needs to take into account that CPU set might have been set
+// to limit the number of cores, or a direct limit on the number of cores.
+// Goal is to select the most restrictive limit.
+func (grp *Group) GetCorrectedCPUCount() int {
+	cpuCount := runtimeNumCPU()
+	cpuSetCount := len(grp.GetCPUSetQuota())
+	if cpuSetCount != 0 && cpuSetCount < cpuCount {
+		cpuCount = cpuSetCount
+	}
+
+	// The reason we are not checking parent groups for any core count limit is that
+	// we want the core count specifically for this group (but restricted to factors of parent group or runtime cpu count).
```

Comment:
line wrap, let's keep it under reasonable length

Reply:
Good point, thanks